### PR TITLE
Add create or update

### PR DESCRIPTION
**Pull request by jmramos02**
/cc @zendesk/mintegrations  @miogalang @joseconsador 

### Description

Add create and update trait

### References
* JIRA: https://zendesk.atlassian.net/browse/MI-639

### Risks
* None

### CODE REVIEW

File: src/Zendesk/API/Resources/Core/Tickets.php [hunk 1]
```diff
@@ -181,6 +181,31 @@ public function createFromTweet(array $params)
...
+     * Create or Update a ticket
+     *
+     * @param array $params
+     *
+     * @throws MissingParametersException
+     * @throws ResponseException
+     * @throws \Exception
+     * @return mixed
+     */
+    public function createOrUpdate(array $params)
```
Comment (miogalang) [inline src/Zendesk/API/Resources/Core/Tickets.php]: Is it not possible to re-use the trait here?

**Comment by jmramos02** [comment]
@miogalang will remove that and use the trait createOrUpdate implementation

File: src/Zendesk/API/Traits/Resource/CreateOrUpdate.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Zendesk\API\Traits\Resource;
+
+use Zendesk\API\Exceptions\RouteException;
+
+trait CreateOrUpdate
```
Comment (miogalang) [inline src/Zendesk/API/Traits/Resource/CreateOrUpdate.php]: Is this only available for tickets? what else has this?

**Comment by jmramos02** [comment]
Users has it, https://developer.zendesk.com/rest_api/docs/core/users#create-or-update-user

**Comment by miogalang** [comment]
@jmramos02 Would be great to have it in all resources that have it. Then you can write a generic unit test for the trait once.

**Comment by jmramos02** [comment]
Okay noted. will rewrite the ticket implementation

**Comment by jmramos02** [comment]
Correction. tickets does not implement create_or_update https://github.com/zendesk/zendesk_api_client_php/issues/241 refers to "users" and not "tickets"

**Comment by miogalang** [comment]
So the create_or_update method is only available for users and nothing else? is that correct?@jmramos02

**Comment by jmramos02** [comment]
Yes @miogalang

**Comment by miogalang** [comment]
@jmramos02 I don't see the point in pulling it out into a trait if it is only used once. We should just keep it in users if it is only available for one. Then think of pulling it out in a trait if we will re use it. This feels like unnecessary optimisation.

**Comment by jmramos02** [comment]
noted @miogalang

File: src/Zendesk/API/Resources/Core/Users.php [hunk 3]
```diff
@@ -9,6 +9,7 @@
 use Zendesk\API\Resources\ResourceAbstract;
 use Zendesk\API\Traits\Resource\CreateMany;
 use Zendesk\API\Traits\Resource\CreateOrUpdateMany;
+use Zendesk\API\Traits\Resource\CreateOrUpdate;
```
Comment (miogalang) [inline src/Zendesk/API/Resources/Core/Users.php]: I think we don't need this anymore

File: tests/Zendesk/API/UnitTests/Core/UsersTest.php [hunk 4]
```diff
@@ -169,4 +169,17 @@ public function testChangePassword()
             $this->client->users($userId)->changePassword($postFields);
         }, "users/{$userId}/password.json", 'PUT');
     }
+
+    /*
+     * Tests if the createOrUpdate function calls the correct endpoint and passes the correct POST data
+     */
+    public function testCreateOrUpdate()
+    {
+        $postFields = ['id' => '1859295', 'name' => 'John Smith'];
```
Comment (miogalang) [inline tests/Zendesk/API/UnitTests/Core/UsersTest.php]: Please use Faker for these values

**Comment by miogalang** [comment]
:+1: good work

**Comment by joseconsador** [comment]
👍

**Comment by pdeuter** [comment]
:chicken:

**Pull request merged**